src/resolve/resolved-dnstls-openssl.c

```diff
@@ -0,0 +1,329 @@
+/* SPDX-License-Identifier: LGPL-2.1+ */
+
+#if !ENABLE_DNS_OVER_TLS || !HAVE_OPENSSL
+#error This source file requires DNS-over-TLS to be enabled and OpenSSL to be available.
+#endif
+
+#include "resolved-dnstls.h"
+#include "resolved-dns-stream.h"
+
+#include <openssl/bio.h>
+#include <openssl/err.h>
+
+DEFINE_TRIVIAL_CLEANUP_FUNC(SSL*, SSL_free);
+DEFINE_TRIVIAL_CLEANUP_FUNC(BIO*, BIO_free);
+
+static int dnstls_flush_write_buffer(DnsStream *stream) {
+        ssize_t ss;
+
+        assert(stream);
+        assert(stream->encrypted);
+
+        if (stream->dnstls_data.write_buffer->length > 0) {
```

Comment:
I do not know it is a right way, but by adding the condition `stream->dnstls_data.handshake > 0` here, then everything seems fine for DNS servers which do not support DNS-over-TLS.

Reply:
Adding a condition for the handshake should theoretically stop resolved from sending the handshake to server. However the issue has been fixed by correcting some error codes as errors during writes resulted in `EPERM` or `EIO` instead of `ECONNREFUSED` or `EPIPE` that are handled correctly.